client/src/api/client.ts

```diff
@@ -0,0 +1,540 @@
+import { errors } from '../util/errors';
+import type * as Types from './components';
+import { operationsByTag } from './components';
+import type { FetcherExtraProps, FetchImpl } from './fetcher';
+import { ApiProviders, getApiUrl } from './providers';
+import type * as Responses from './responses';
+import type * as Schemas from './schemas';
+
+export interface XataApiClientOptions {
+  fetchImpl: FetchImpl;
+  apiKey: string;
+  api?: ApiProviders;
+}
+
+export class XataApiClient {
+  private extraProps: FetcherExtraProps;
+
+  constructor(options: XataApiClientOptions) {
+    const fetchImpl = typeof fetch !== 'undefined' ? fetch : options.fetchImpl;
+    if (!fetchImpl) {
+      throw new Error(errors.noFetchImplementation);
+    }
+
+    const provider = options.api ?? 'production';
+
+    this.extraProps = {
+      apiUrl: getApiUrl(provider, 'main'),
+      workspacesApiUrl: getApiUrl(provider, 'workspaces'),
+      fetchImpl,
+      apiKey: options.apiKey
+    };
+  }
+
+  public get user() {
+    return new UserApi(this.extraProps);
+  }
+
+  public get workspace() {
+    return new WorkspaceApi(this.extraProps);
+  }
+
+  public get database() {
+    return new DatabaseApi(this.extraProps);
+  }
+
+  public get branch() {
+    return new BranchApi(this.extraProps);
+  }
+
+  public get table() {
+    return new TableApi(this.extraProps);
+  }
+
+  public get records() {
+    return new RecordsApi(this.extraProps);
+  }
+}
+
+class UserApi {
+  constructor(private extraProps: FetcherExtraProps) {}
+
+  public getUser(): Promise<Schemas.UserWithID> {
+    return operationsByTag.users.getUser({ ...this.extraProps });
```

Comment:
This is not auto-generated, right? How hard would it be to generate these in an automated way?

Reply:
I thought of generating this, but I liked that we get to manually decide which changes we want to introduce to the public facing SDK API. ``tsc`` will type check that the wrapper matches the types from the ``auto-generated`` functions.